internal/clients/http/ec2/iam_client.go

```diff
@@ -0,0 +1,304 @@
+			"ec2:CreateLaunchTemplateVersion",
+			"ec2:CreateTags",
+			"ec2:DeleteKeyPair",
+			"ec2:DeleteTags",
+			"ec2:DescribeAvailabilityZones",
+			"ec2:DescribeImages",
+			"ec2:DescribeInstanceTypes",
+			"ec2:DescribeInstances",
+			"ec2:DescribeKeyPairs",
+			"ec2:DescribeLaunchTemplates",
+			"ec2:DescribeLaunchTemplateVersions",
+			"ec2:DescribeRegions",
+			"ec2:DescribeSecurityGroups",
+			"ec2:DescribeSnapshotAttribute",
+			"ec2:DescribeTags",
+			"ec2:ImportKeyPair",
+			"ec2:RunInstances",
+			"ec2:StartInstances",
+			"iam:ListRolePolicies",
+		},
+	}
+}
+
+func getRoleName(arn string) (string, error) {
+	arnParts := strings.Split(arn, ":")
+	if len(arnParts) == 0 {
+		return "", fmt.Errorf("%w: ARN has no colons: %s", http.ARNParsingError, arn)
+	}
+	roleName := strings.Split(arnParts[len(arnParts)-1], "/")
+	if len(roleName) != 2 {
+		return "", fmt.Errorf("%w: ARN has incorrect syntax: %s rolename parsing result: %s",
+			http.ARNParsingError, arn, roleName)
+	} else if roleName[0] != "role" {
+		return "", fmt.Errorf("%w: ARN does not have any role: %s", http.ARNParsingError, arn)
+	}
+	return roleName[1], nil
+}
+
+func getPermissionsFromAllowed(ctx context.Context, statementBody interface{}) []string {
+	switch statementFields := statementBody.(type) {
+	case map[string]interface{}:
+		switch effectField := statementFields["Effect"].(type) {
+		case string:
+			if effectField == "Allow" {
+				return getPermissionsFromStatement(ctx, statementFields)
+			}
+		}
+	}
+	return nil
+}
+
+func getPermissionsFromStatement(ctx context.Context, statementBody map[string]interface{}) []string {
+	var result []string
+
+	switch actionField := statementBody["Action"].(type) {
+	case string:
+		result = append(result, actionField)
+	case []interface{}:
+		for _, v := range actionField {
+			switch actionElement := v.(type) {
+			case string:
+				result = append(result, actionElement)
+			}
+		}
+	}
+	return result
+}
+
+func getJsonFromAWSDocument(ctx context.Context, document string) (string, error) {
+	unescapedDoc, err := url.QueryUnescape(document)
+	if err != nil {
+		return "", fmt.Errorf("could not parse URL from the document: %w", err)
+	}
+	return unescapedDoc, nil
+}
+
+func getStatementFromJson(ctx context.Context, document string) ([]string, error) {
+	var result []string
+
+	jsonBody := []byte(document)
+	var policy interface{}
+	err := json.Unmarshal(jsonBody, &policy)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal json: %w", err)
+	}
+
+	switch policyBody := policy.(type) {
+	case map[string]interface{}:
+		switch statementField := policyBody["Statement"].(type) {
+		case []interface{}:
+			for _, v := range statementField {
+				switch statementElement := v.(type) {
+				case interface{}:
+					result = append(result, getPermissionsFromAllowed(ctx, statementElement)...)
+				}
+			}
+		case interface{}:
+			result = append(result, getPermissionsFromAllowed(ctx, statementField)...)
+
+		}
+	}
```

Comment:
Why do you do this empty interface parsing? You should be able to unmarshal directly into a type, Go JSON parser is able to ignore unknown fields AFAIK and the format [seems to be documented](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements.html).

Reply:
Okay then this is indeed correct.